services/src/main/java/com/broadleafcommerce/subscriptionoperation/service/DefaultSubscriptionPricingService.java

```diff
@@ -0,0 +1,875 @@
+     * applicable. The assumption is that this date is sometime in the future.
+     *
+     * @param subscriptionRootItem The {@link CartItem} that represents the root of the subscription
+     *        items
+     * @param pricingContext The {@link SubscriptionPricingContext} in which this operation is
+     *        occurring
+     * @param contextInfo context surrounding the multi-tenant state
+     * @return a non-typical next bill date for the subscription or null if one is not applicable.
+     */
+    protected Instant determineAtypicalNextBillDate(@lombok.NonNull CartItem subscriptionRootItem,
+            @lombok.NonNull SubscriptionPricingContext pricingContext,
+            @Nullable ContextInfo contextInfo) {
+        return null;
+    }
+
+    protected Map<Integer, PeriodDefinition> buildPeriodDefinitions(
+            @lombok.NonNull CartItem subscriptionRootItem,
+            @lombok.NonNull SubscriptionPricingContext pricingContext,
+            @Nullable ContextInfo contextInfo) {
+        Map<Integer, PeriodDefinition> estimatedFuturePaymentPeriods = new HashMap<>();
+
+        Instant previousBillDate = Optional.ofNullable(pricingContext.getAtypicalNextBillDate())
+                .map(atypicalNextBillDate -> atypicalNextBillDate.truncatedTo(ChronoUnit.DAYS))
+                .orElse(pricingContext.getSubscriptionActionDate().truncatedTo(ChronoUnit.DAYS));
+
+        int estimatedFuturePaymentCount =
+                getEstimatedFuturePaymentCount(subscriptionRootItem, pricingContext, contextInfo);
+        for (int period = 1; period <= estimatedFuturePaymentCount; period++) {
+            if (period == 1 && pricingContext.getAtypicalNextBillDate() != null) {
+                PeriodDefinition periodDefinition = buildAtypicalFirstPeriodDefinition(
+                        estimatedFuturePaymentPeriods, period, pricingContext, contextInfo);
+
+                previousBillDate = periodDefinition.getBillDate();
+                continue;
+            }
+
+            PeriodDefinition periodDefinition = typeFactory.get(PeriodDefinition.class);
+            Instant billDate =
+                    addOnePeriodToDate(previousBillDate, pricingContext, contextInfo);
+            periodDefinition.setBillDate(billDate);
+
+            Instant periodStartDate =
+                    determinePeriodStartDate(previousBillDate, billDate, pricingContext,
+                            contextInfo);
+            periodDefinition.setPeriodStartDate(periodStartDate);
+
+            Instant periodEndDate =
+                    determinePeriodEndDate(period, periodDefinition, pricingContext, contextInfo);
+            periodDefinition.setPeriodEndDate(periodEndDate);
+
+            estimatedFuturePaymentPeriods.put(period, periodDefinition);
+            previousBillDate = periodDefinition.getBillDate();
+        }
+
+        return estimatedFuturePaymentPeriods;
+    }
+
+    protected PeriodDefinition buildAtypicalFirstPeriodDefinition(
```

Comment:
Seems more polite to say custom than atypical.
```suggestion
    protected PeriodDefinition buildCustomFirstPeriodDefinition(
```

Reply:
Planning to leave this one as-is